### PR TITLE
Add and use smarter HIR builders

### PR DESCRIPTION
**Pull request by frankmcsherry**
Following a thought from @mjibson, I added a few that were missing here but present in MIR, and introduced a bit more use.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by benesch — APPROVED** [review]
The Kimballizer!

**Comment by mjibson** [comment]
For history: all I did was post a link in slack. Original idea from https://buttondown.email/jaffray/archive/simplifying-expressions-bottom-up/

**Review by ParkMyCar — APPROVED** [review]
Neat!

File: src/sql/src/plan/expr.rs [hunk 1]
```diff
@@ -1329,10 +1329,21 @@ impl HirRelationExpr {
...
+        {
+            predicates.extend(preds);
+            predicates.sort();
+            predicates.dedup();
+            self
+        } else {
+            HirRelationExpr::Filter {
+                input: Box::new(self),
+                predicates: preds,
+            }
```
Comment (ggevay) [inline src/sql/src/plan/expr.rs:1346]: Maybe sort and dedup also in the else branch?

**Review by ggevay — APPROVED** [review]
Looks good, just one minor comment.

**Comment by aalexandrov** [comment]
FYI, Tiark Rompf calls these "smart constructors" [in his thesis](https://infoscience.epfl.ch/record/180642?ln=en), but doesn't provide a PL reference. I'm pretty sure this is an FP idea, though, so maybe @mgree might know more about those.

**Comment by mgree** [comment]
Yes, these are smart constructors. I think they were first used in Lisp, forever ago, but I don't have a reference. The term is usually used uncited in the FP literature.

**Pull request merged**